examples/interactive/cli-platform/Arg.roc

```diff
@@ -0,0 +1,782 @@
+interface Arg
+    exposes [
+        Parser,
+        NamedParser,
+        parse,
+        toHelp,
+        parseFormatted,
+        succeed,
+        bool,
+        str,
+        subCommand,
+        choice,
+        withParser,
+        program,
+    ]
+    imports []
+
+## A parser for a command-line application.
+## A [NamedParser] is usually built from a [Parser] using [program].
+NamedParser a := {
+    name : Str,
+    help : OptionStr,
+    parser : Parser a,
+}
+
+## Describes how to parse a slice of command-line arguments.
+## [Parser]s can be composed in various ways, including via [withParser] and
+## [subCommand].
+## Once you have a [Parser] that describes your application's entire parsing
+## needs, consider transforming it into a [NamedParser].
+Parser a := [
+    Succeed a,
+    Arg Config (List Str -> Result a [NotFound, WrongType]),
+    # TODO: hiding the record behind an alias currently causes a panic
+    SubCommand
+        (List {
+            name : Str,
+            parser : Parser a,
+        }),
+
+    # Constructed during transformations of the above variants
+    WithConfig (Parser a) Config,
+    Lazy ({} -> a),
+]
+
+## Enumerates errors that can occur during parsing a list of command line arguments.
+ParseError a : [
+    ## The program name was not found as the first argument to be parsed.
+    ProgramNameNotProvided Str,
+    ## An argument is required, but it was not found.
+    MissingRequiredArg Str,
+    ## An argument was found, but it didn't have the expected [Type].
+    WrongType
+        {
+            arg : Str,
+            expected : Type,
+        },
+    ## A subcommand is required, but it was not found.
+    SubCommandNotFound
+        {
+            choices : List Str,
+        },
+    ## A subcommand was found, but it was not the expected one.
+    IncorrectSubCommand
+        {
+            found : Str,
+            choices : List Str,
+        },
+]a
+
+## Expected type of an argument, in an argument list being parsed.
+## Describes how a string argument should be interpreted as a certain type.
+Type : [
+    Str,
+    Bool,
+]
+
+## Help metadata extracted from a [Parser].
+Help : [
+    SubCommands (List { name : Str, help : Help }),
+    Config (List Config),
+]
+
+OptionStr : [Some Str, NotProvided]
+
+Config : {
+    long : Str,
+    short : OptionStr,
+    help : OptionStr,
+    type : Type,
+}
+
+# TODO currently unused because using this as an annotation yields panics
+# OptConfig : {
+#     long : Str,
+#     short ?OptionStr,
```

Comment:
I was thinking we could do this as just `Str` instead of `OptionStr` and treat empty string as equivalent to current `None`

That makes it nicer to call (you don't need to say `Some` every time you're actually specifying it) and since in both the case of `short` and `help` I don't think there's a meaningful distinction between passing `Some ""` and `NotProvided` anyway!